### PR TITLE
Re-implement JObject into c-style struct

### PR DESCRIPTION
**Pull request by yichoi**
Related issue: #439

* Reference counting
  - `iotjs_jval_t`  : Increase jerry refcount. This type of variable must be destroyed by `iotjs_jval_destroy()`
  - `iotjs_jval_t*` : This type of variable should not be destroyed.

* Added `valgrind_tracer` in `IOTJS_VALIDATED_STRUCT`.
  - To find out missing iotjs_jval_destroy() call
  - Allocate small memory on create(), and release it on destroy(), so if someone does not call destroy(), it will report where that memory was allocated at the first place.
  - With only jerry assertion (e.g. Assertion 'ECMA_STRING_IS_REF_EQUALS_TO_ONE (string_p)' failed), it was hard to find out where exactly jerry_value_release() is missing. (Help wanted if there're some better ways!)

### CODE REVIEW

File: src/iotjs.cpp [hunk 1]
```diff
@@ -95,50 +95,47 @@ static void CleanupModules() {
...
+    UncaughtException(&jmain_res);
   }
+  iotjs_jval_destroy(&jmain_res);
+
+  return !throws;
 }
 
 
 static bool StartIoTjs(Environment* env) {
   // Initialize jerry null and undefined objects.
```
Comment (chokobole) [inline src/iotjs.cpp]: remove this comment asyou remove Jobject::init()

File: src/iotjs.cpp [hunk 2]
```diff
@@ -167,7 +164,6 @@ static bool StartIoTjs(Environment* env) {
   CleanupModules();
 
   // Release jerry null and undefined objects.
```
Comment (chokobole) [inline src/iotjs.cpp]: ditto

File: src/iotjs_binding.cpp [hunk 3]
```diff
@@ -19,538 +19,459 @@
...
+  iotjs_jval_set_property_jval(jobj, name, &jval);
+  iotjs_jval_destroy(&jval);
+}
 
-  char* buffer = iotjs_buffer_allocate(size + 1);
-  jerry_char_t* jerry_buffer = reinterpret_cast<jerry_char_t*>(buffer);
 
-  size_t check = jerry_string_to_char_buffer(_obj_val, jerry_buffer, size);
+iotjs_jval_t iotjs_jval_get_property(const iotjs_jval_t* jobj,
+                                            const char* name) {
```
Comment (chokobole) [inline src/iotjs_binding.cpp]: over indented

File: src/iotjs_binding.cpp [hunk 4]
```diff
@@ -587,235 +513,253 @@ void iotjs_jargs_destroy(iotjs_jargs_t* jargs) {
...
-  delete _this->argv[i];
-  _this->argv[i] = new iotjs::JObject(*x);
+  iotjs_jval_destroy(&_this->argv[index]);
+  _this->argv[index] = iotjs_jval_create_copied(x);
 }
 
 
-iotjs::JObject* iotjs_jargs_get(iotjs_jargs_t* jargs, uint16_t i) {
-  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_jargs_t, jargs);
+const iotjs_jval_t* iotjs_jargs_get(const iotjs_jargs_t* jargs, uint16_t index){
```
Comment (chokobole) [inline src/iotjs_binding.cpp]: space between ) {

File: src/iotjs_handlewrap.h [hunk 5]
```diff
@@ -43,8 +43,8 @@ namespace iotjs {
 // The javascript object now can be reclaimed by GC.
 class HandleWrap : public JObjectWrap {
  public:
-  HandleWrap(JObject& jobject, /* Object that connect with the uv handle*/
-             uv_handle_t* handle);
+  HandleWrap(const iotjs_jval_t* jobject, uv_handle_t* handle);
+  // jobject: Object that connect with the uv handle*/
```
Comment (chokobole) [inline src/iotjs_handlewrap.h]: needless */

File: src/module/iotjs_module_buffer.h [hunk 6]
```diff
@@ -23,24 +23,24 @@
 namespace iotjs {
 
 
-JObject* InitBuffer();
+iotjs_jval_t InitBuffer();
 
 
 // Crate buffer object.
```
Comment (chokobole) [inline src/module/iotjs_module_buffer.h]: Create

File: src/module/iotjs_module_tcp.cpp [hunk 7]
```diff
@@ -137,15 +138,15 @@ static void AfterConnect(uv_connect_t* req, int status) {
 
   // Take callback function object.
   //  function afterConnect(status)
```
Comment (chokobole) [inline src/module/iotjs_module_tcp.cpp]: over space

File: src/module/iotjs_module_timer.h [hunk 8]
```diff
@@ -21,15 +21,15 @@
...
+iotjs_jval_t InitTimer();
 
 class TimerWrap : public HandleWrap {
  public:
-  explicit TimerWrap(Environment* env, JObject& jtimer)
-      : HandleWrap(jtimer, reinterpret_cast<uv_handle_t*>(&_handle))
-      , _jcallback(NULL) {
+  explicit TimerWrap(Environment* env, const iotjs_jval_t* jtimer)
+      : HandleWrap(jtimer, reinterpret_cast<uv_handle_t*>(&_handle)) {
     // Initialze timer handler.
```
Comment (chokobole) [inline src/module/iotjs_module_timer.h]: Initialize

**Comment by yichoi** [comment]
@chokobole, Thank you for the review. I've updated the PR as you commented.

And I've added one more modification: rename iotjs_binding.cpp into iotjs_binding.c.

File: src/iotjs.cpp (hunk 1 above)
Comment (yichoi) [inline src/iotjs.cpp]: In fact, `iotjs_binding_initialize` does almost the same thing.
I've updated this comment to include more values that `iotjs_binding_initialize` initializes.

File: src/iotjs_binding.h [hunk 9]
```diff
@@ -26,309 +26,286 @@
...
-  JObject& value();
-  JResultType type() const;
-
-  bool IsOk() const;
-  bool IsException() const;
-
- private:
-  JObject _value;
-  JResultType _type;
+struct iotjs_jval_t;
```
Comment (LaszloLango) [inline src/iotjs_binding.h]: A cannot find where is `iotjs_jval_t` defined

File: src/iotjs.cpp [hunk 10]
```diff
@@ -166,8 +163,7 @@ static bool StartIoTjs(Environment* env) {
   // Release bulitin modules.
   CleanupModules();
 
-  // Release jerry null and undefined objects.
-  JObject::cleanup();
+  // Release commonly used jerry values
```
Comment (zherczeg) [inline src/iotjs.cpp]: I noticed that the original comments terminated by dot (.), but the new comments does not. Is this an intentional change?

File: src/iotjs_binding.h (hunk 9 above)
Comment (yichoi) [inline src/iotjs_binding.h]: This is useless line. iotjs_jval_t is defined in line 57-59, like below
```
typedef struct {
  jerry_value_t value;
} IOTJS_VALIDATED_STRUCT(iotjs_jval_t);
```

File: src/iotjs.cpp (hunk 10 above)
Comment (yichoi) [inline src/iotjs.cpp]: No it was not intentional change, I've fixed it while rebasing the commit.

File: src/module/iotjs_module_fs.cpp [hunk 11]
```diff
@@ -63,24 +67,26 @@ static void After(uv_fs_t* req) {
...
-        iotjs_jargs_append_obj(&jarg, &ret);
+        iotjs_jval_t ret = MakeStatObject(&s);
+        iotjs_jargs_append_jval(&jarg, &ret);
+        iotjs_jval_destroy(&ret);
         break;
       }
-      default:
+      default: {
         iotjs_jargs_append_null(&jarg);
+      }
```
Comment (zherczeg) [inline src/module/iotjs_module_fs.cpp]: I would add a break here.

File: src/module/iotjs_module_fs.cpp [hunk 12]
```diff
@@ -196,42 +197,37 @@ JHANDLER_FUNCTION(Read) {
...
+  BufferWrap* buffer_wrap = BufferWrap::FromJBuffer(jbuffer);
   char* data = buffer_wrap->buffer();
   int data_length = buffer_wrap->length();
   JHANDLER_CHECK(data != NULL);
   JHANDLER_CHECK(data_length > 0);
 
   if (offset >= data_length) {
-    JHANDLER_THROW_RETURN(RangeError, "offset out of bound");
+    JHANDLER_THROW(RANGE, "offset out of bound");
   }
```
Comment (zherczeg) [inline src/module/iotjs_module_fs.cpp]: Is it a good idea if the second throw overwrites the first throw? Isn't it create a memory leak?

File: src/module/iotjs_module_httpparser.cpp [hunk 13]
```diff
@@ -308,64 +305,64 @@ void HTTPParserWrap::Initialize(http_parser_type type) {
 }
 
 
+static HTTPParserWrap* getParserWrap(const iotjs_jval_t* jparser) {
+  uintptr_t handle = iotjs_jval_get_object_native_handle(jparser);
+  return reinterpret_cast<HTTPParserWrap*>(handle);
+}
+
```
Comment (zherczeg) [inline src/module/iotjs_module_httpparser.cpp]: I think double newline should be used here

**Comment by zherczeg** [comment]
Great patch!

File: src/module/iotjs_module_httpparser.cpp (hunk 13 above)
Comment (yichoi) [inline src/module/iotjs_module_httpparser.cpp]: I added one more newline here.

**Comment by yichoi** [comment]
Updated the commit after reviews.

**Comment by zherczeg** [comment]
LGTM

**Pull request merged**